src/lj_cparse.c

```diff
@@ -1610,10 +1612,20 @@ static void cp_decl_array(CPState *cp, CPDecl *decl)
   CTInfo info = CTINFO(CT_ARRAY, 0);
   CTSize nelem = CTSIZE_INVALID;  /* Default size for a[] or a[?]. */
   cp_decl_attributes(cp, decl);
-  if (cp_opt(cp, '?'))
+  if (cp_opt(cp, '?')) {
     info |= CTF_VLA;  /* Create variable-length array a[?]. */
-  else if (cp->tok != ']')
-    nelem = cp_expr_ksize(cp);
+  } else if (cp->tok != ']') {
+    CPValue k;
+    CType *ct;
+    cp_expr_sub(cp, &k, 0);
+    ct = ctype_raw(cp->cts, k.id);
+    if (ctype_isinteger(ct->info)) {
+      if (k.u32 < 0x80000000u)
```

Comment:
I realized that this works by accident. `ctype_isinteger()` does not guarantee that the result is a compile-time constant, so `k.u32` may be the parameter index. Currently the code does not seem to be tracking whether a `CPValue` is a compile-time constant. I will add this and post an update.